src/service.rs

```diff
@@ -188,59 +203,69 @@ impl ServiceStore {
             }
 
             // Remove the endpoint from the service.
-            if let Some(wls) = self.by_vip.get_mut(vip) {
-                wls.write().unwrap().endpoints.remove(addr);
+            if let Some(service) = self.by_vip.get(vip) {
+                // Clone the service and remove the endpoint.
+                let mut service = service.deref().clone();
+                service.endpoints.remove(addr);
+
+                // Update the service.
+                self.insert(service);
             }
         }
     }
 
     /// Adds the given service.
     pub(super) fn insert(&mut self, mut service: Service) {
         // First mutate the service and add all missing endpoints
-        for addr in service.addresses.as_slice() {
+        for addr in &service.vips {
             // Due to ordering issues, we may have gotten workloads with VIPs before we got the service
             // we should add those workloads to the vips map now
-            if let Some(wips_to_endpoints) = self.staged_vips.remove(addr) {
-                for (wip, ep) in wips_to_endpoints {
-                    service.endpoints.insert(wip.clone(), ep);
-                }
-            }
-
-            // If svc already exists, add old endpoints to new svc
-            if let Some(prev) = self.by_vip.get_mut(addr) {
-                let prev = prev.read().unwrap();
-                for (wip, ep) in prev.endpoints.iter() {
-                    service.endpoints.insert(wip.clone(), ep.clone());
+            if let Some(endpoints) = self.staged_vips.remove(addr) {
+                for (wip, ep) in endpoints {
+                    // Don't override if the new service has the endpoint.
+                    if !service.endpoints.contains_key(&wip) {
+                        service.endpoints.insert(wip.clone(), ep);
+                    }
```

Comment:
this looks like a change in behavior.. was this a bug? is there a test to cover this case?

Reply:
The previous code relied on the fact that all inserted `Services` had no endpoints. This was because it was called with services derived from the raw XDS, which don't contain endpoints.

This is no longer the case, however, due to the fact that services are now immutable as they are stored as `Arc<Service>`. Now when a service is updated, we have to remove it and then add the updated one in its place.

Existing tests already cover this case.